lessons/advanced/gen-stage.md

```diff
@@ -0,0 +1,261 @@
+
+We'll begin by generating a project with a supervision tree:
+
+```shell
+$ mix new genstage_example --sup
+$ cd genstage_example
+```
+
+Let's update our dependencies in `mix.exs` to include `gen_stage`:
+
+```elixir
+  defp deps do
+    [
+      {:gen_stage, "~> 0.7},
+    ]
+  end
+```
+
+We should fetch our dependencies and compile before go much further:
+
+```shell
+$ mix do deps.get, compile
+```
+
+Now we're ready to build our producer!
+
+## Producer
+
+The first step of our GenStage application is creating our producer.  As we discussed before we want to create a producer that emits a constant stream of numbers.  Let's create our producer file:
+
+```shell
+$ mkdir lib/genstage_example
+$ touch lib/genstage_example/producer.ex
+```
+
+Now we can add the code:
+
+```
+defmodule GenstageExample.Producer do
+  alias Experimental.GenStage
+  
+  use GenStage
+  
+  def start_link(initial \\ 0) do
+    GenStage.start_link(__MODULE__, initial, name: __MODULE__)
+  end
+
+  def init(counter), do: {:producer, counter}
+
+  def handle_demand(demand, state) do
+    events = Enum.to_list(state..state + demand - 1)
+    {:noreply, events, (state + demand)}
+  end
+end
+```
+
+The two most important parts to take note of here are `init/1` and `handle_demand/2`.  In `init/1` we set the initial state as we've done in our GenServers but more importantly we label ourselves as a producer.  The response from our `init/1` function is what GenStage relies upon to classify our process.
+
+The `handle_demand/2` function is the real meat of our producer and must be implemented by all GenStage producers.  Here we return the set of numbers demanded by our consumers and increment our counter.
+
+## Producer Consumer
+
+Now that we have have a number generating producer let's move on to our producer-consumer.  We'll want to request numbers from our producer, filter out the odd one, and respond to demand.
+
+```shell
+$ touch lib/genstage_example/producer_consumer.ex
+```
+
+Let's update our file to look like the example code:
+
+```
+defmodule GenstageExample.ProducerConsumer  do
+  alias Experimental.GenStage
+  use GenStage
+  
+  require Integer
+  
+  def start_link do
+    GenStage.start_link(__MODULE__, :state_doesnt_matter, name: __MODULE__)
+  end
+  
+  def init(state) do
+    {:producer_consumer, state, subscribe_to: [GenstageExample.Producer]}
+  end
+
+  def handle_events(events, _from, state) do
+    numbers = 
+    	events 
+    	|> Enum.filter(&Integer.is_even/1)
+    	
+    {:noreply, numbers, state}
+  end
+end
+```
+
+You may have noticed with our producer-consumer we've introduced a new function `handle_events/3` and option in `init/1`.  With the `subscribe_to` option we instruct GenStage to put us into communcation with a specific producer.
```

Comment:
It might be worth mentioning that the use of `handle_events/3` is specifically what makes it a producer-consumer per the documentation